crates/engine/tree/src/tree/mod.rs

```diff
@@ -1134,6 +1134,50 @@ where
         Ok(InsertPayloadOk::Inserted(BlockStatus::Valid(attachment)))
     }
 
+    /// Updates the tracked finalized block if we have it.
+    fn update_finalized_block(&self, finalized_block_hash: B256) {
+        if finalized_block_hash.is_zero() {}
+
+        // TODO find finalized block and ensure it's canonical
+        // TODO tree cleanup
+    }
+
+    /// Updates the tracked safe block if we have it
+    fn update_safe_block(&self, safe_block_hash: B256) {
+        if safe_block_hash.is_zero() {}
+
+        // TODO find safe block and ensure it's canonical
+    }
+
+    /// Ensures that the given forkchoice state is consistent, assuming the head block has been
+    /// made canonical.
+    ///
+    /// If the forkchoice state is consistent, this will return Ok(()). Otherwise, this will
+    /// return an instance of [`OnForkChoiceUpdated`] that is INVALID.
+    ///
+    /// This also updates the safe and finalized blocks in the [`CanonicalInMemoryState`], if they
+    /// are consistent with the head block.
+    fn ensure_consistent_forkchoice_state(
+        &self,
+        state: ForkchoiceState,
+    ) -> Result<(), OnForkChoiceUpdated> {
+        // Ensure that the finalized block, if not zero, is known and in the canonical chain
+        // after the head block is canonicalized.
+        //
+        // This ensures that the finalized block is consistent with the head block, i.e. the
+        // finalized block is an ancestor of the head block.
+        self.update_finalized_block(state.finalized_block_hash);
```

Comment:
thinking that the `update_` methods should return `Result`s in case they are not consistent, if we're going to be doing validation there

Reply:
yeah will add this in the next step